api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java

```diff
@@ -479,13 +481,26 @@ public void assignBillingProject_locking() throws InterruptedException, Executio
   @Test
   public void cleanBillingBuffer_creating() {
     DbBillingProjectBufferEntry entry = new DbBillingProjectBufferEntry();
+    entry.setFireCloudProjectName("foo");
     entry.setStatusEnum(CREATING, this::getCurrentTimestamp);
     entry.setCreationTime(getCurrentTimestamp());
     billingProjectBufferEntryDao.save(entry);
 
+    doReturn(new BillingProjectStatus().creationStatus(CreationStatusEnum.CREATING))
+        .when(fireCloudService)
+        .getBillingProjectStatus(entry.getFireCloudProjectName());
+    billingProjectBufferService.syncBillingProjectStatus();
+
     CLOCK.setInstant(NOW.plus(61, ChronoUnit.MINUTES));
     billingProjectBufferService.cleanBillingBuffer();
 
+    // Time has elapsed but no status sync has occurred, should not transition.
+    assertThat(billingProjectBufferEntryDao.findOne(entry.getId()).getStatusEnum())
+        .isEqualTo(CREATING);
+
+    // Sync again, now @ +61 minutes - should transition.
+    billingProjectBufferService.syncBillingProjectStatus();
```

Comment:
Should we always call sync before clean? If so, why not have the latter call the former for us? 

I'm starting to think this service may be exhibiting Temporal Coupling. It's hard to understand what the invariants are w/r/t the billing buffer.

Reply:
sync processes a limited number of entries on every run. Yes, it was exhibiting Temporal Coupling, though probably not intentionally. That is the bug this PR is fixing. I'm not aware of any more instances of that pattern in the current buffer system.

Here is the original design doc: http://broad.io/1ppw, it doesn't contain any details about the different cronjobs. Brief summary which we should probably writeup as a documentation task:

Crons:
bufferBillingProjects (every minute): create new entries in the buffer, always in the CREATING state
syncBillingProjectStatus (every minute): poll for state changes on the 5 least recently synced entries in a relevant state, namely tracking transitions from CREATING --> ACTIVE or ERROR
cleanBillingBuffer (every hour): clean bad entries out of the buffer, e.g. things that have been in a particular state for too long; this is necessary as the buffer has a fixed capacity

Methods (buffer consumer):
create/clone workspace: mark an entry as ASSIGNING, then ASSIGNED when used for a workspace. If we fail in the middle, cleanBillingBuffer cleans out entries stuck in ASSIGNING

Reply:
Awesome, thanks. That's helpful. One thing that jumps out is, why don't we consolidate sync with clean, since both are concerned with updating status based on elapsed time.

Reply:
In theory you could combine all three of these cron jobs. They all have different roles, and I think it is a better design to allow these pieces to operate independently both for failure isolation, Firecloud load, and to allow entries to move quickly through the state machine. cleanBillingBuffer is also not scoped exclusively to Firecloud project status changes